File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.LinkedList;
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> list = new LinkedList<>();
        Deque<TreeNode> queue = new LinkedList<>();
        boolean left = true;
        int level = 0;
        int nulls = 0;
        
        queue.add(root);
        
        while(queue.size() > 0){
            list.add(new LinkedList<>());    
            int len = queue.size();
            
            for(int i = 0; i < len; i++){
                if(left){
                    TreeNode node = queue.removeLast();
                    if(node != null){
                        list.get(level).add(node.val);
                        queue.addFirst(node.left);
                        queue.addFirst(node.right);
                    }else{
                        nulls++;
                    }
                }else{
                   TreeNode node = queue.pollFirst();  
                    if(node != null){
                        list.get(level).add(node.val);
                        queue.addLast(node.right);
                        queue.addLast(node.left);
                    }else{
                        nulls++;
                    }
                }
            }
            if(nulls == len){list.remove(level);}
            left = !left;
            level++;
            nulls = 0;
        }
        return list;
    }
}

File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        deque<TreeNode*> q;
        q.push_back(root);
        bool left = true;
        vector<int> level(0);
        vector<vector<int>> ans(0);
        while(!q.empty()){
            TreeNode* node = NULL;
            int len = q.size();
            
            for(int i = 0; i < len; i++){

                if(left){
                    node = q.back();
                    q.pop_back();
                    if(node != NULL){
                        q.push_front(node->left);
                        q.push_front(node->right);
                        level.push_back(node->val);
                    }
                }else{
                    node = q.front();
                    q.pop_front();
                    if(node != NULL){
                        q.push_back(node->right);
                        q.push_back(node->left);
                        level.push_back(node->val);
                    }
                }
            }
            
            left = !left;
            if(level.size()>0){
                ans.push_back(level);
            }
            level.clear();
        }
        return ans;
    }
};